# Analysis of Amazon Product Data

After construction of the graph (product-id and link to cross-buys), the following topics will be investigated:
* correlation between in-degrees and salesrank
* and some further smart questions... influence of number of reviews?, are there different sub-graphs like produkt categories?
* 
* 
* 
Data source: https://snap.stanford.edu/data/amazon-meta.html

Data import

In [3]:
# Import graphframes (from Spark-Packages)

from graphframes import *
from pyspark.sql.functions import monotonically_increasing_id, desc


# Get the product data (already pre-processed data in table format)
products = spark.read.csv("/FileStore/tables/ProductData_small.csv", sep=";", header=True)

# The the cross-buy data (similar products)
links = spark.read.csv("/FileStore/tables/LinkData_small.csv", sep=";", header=True)

How many products are we dealing with here?

In [5]:
products.count()

Out[ 2 ]: 10611

quick check of the data formats

In [7]:
display(products.take(3))

Id ASIN Group Salesrank Reviews Title 21 0790747324 DVD 795 140 The Time Machine 84 B000063W82 Video 780 14 The Best of Schoolhouse Rock! - 30th Anniversary Edition 252 B0000262WI Music 581 82 The Köln Concert

And which and how many groups are there?

In [9]:
display(products.select("Group").distinct())

Group Video Toy DVD Sports Baby Product Video Games Book Music Software

In [10]:
products.select("Group").distinct().count()

Out[ 5 ]: 9

And how many products per category?

In [12]:
display(products.groupBy('Group').count().sort('count', ascending=False))

Group count Video 2933 Music 2855 Book 2535 DVD 2276 Software 5 Toy 4 Sports 1 Baby Product 1 Video Games 1

In [13]:
display(links.take(3))

src dest 0790747324 B00007JMD8 0790747324 6305350221 0790747324 B00004RF9B

The graph will be constructed based on the product data. Each product is identified through its unique *ASIN* (Amazon Standard Identification Number) code

GraphFrame library required *ASIN* to be renamed to *id*, but be careful, there is already an "Id" in the DataFrame

In [15]:
# rename "Id" to "Id-nu" to avoid any confusion (nu = not used)
# rename "ASIN" to "id"
ProductVertices = products.withColumnRenamed("Id", "Id-nu").distinct()
ProductVertices = ProductVertices.withColumnRenamed("ASIN", "id").distinct()

# check if done correctly
display(ProductVertices.take(3))

Id-nu id Group Salesrank Reviews Title 13260 B000002TIR Music 2004 16 French Kiss 13683 0140281916 Book 2877 16 Bargaining for Advantage : Negotiation Strategies for Reasonable People 30773 B000001EYO Music 3579 31 The Big Lebowski: Original Motion Picture Soundtrack

How does the Edge table look like?

In [17]:
display(links.take(1))

src dest 0790747324 B00007JMD8

And how many edges are there?

In [19]:
links.count()

Out[ 11 ]: 52158

Although not always strictly required, it's good practice to have an ID column for the tables. 

Since the product links does not have one yet, create and additional ID:

In [21]:
# 2. Create Edges
links = links.withColumn("link-id", monotonically_increasing_id())

display(links.take(3))

src dest link-id 0790747324 B00007JMD8 0 0790747324 6305350221 1 0790747324 B00004RF9B 2

Salesrank and number of reviews are read as string. Change type to int for calculations.

In [23]:
# change type of features "Salesrank" and "Reviews" to int
ProductVertices = ProductVertices.withColumn('Salesrank', ProductVertices.Salesrank.cast('int'))
ProductVertices = ProductVertices.withColumn('Reviews', ProductVertices.Reviews.cast('int'))
display(ProductVertices.take(1))

Id-nu id Group Salesrank Reviews Title 13260 B000002TIR Music 2004 16 French Kiss

Rename links to similar products ar *src* and *dst*

In [25]:
# rename destination "dest" to "dst"
links = links.withColumnRenamed('dest', 'dst')


Now comes the moment of truth! Let's see if the graph can be constructed!

In [27]:
# Build the GraphFrame based on products and links

ProductGraph = GraphFrame(ProductVertices, links)

# print a sample of the vertices and edges
print("Products: %d" % ProductGraph.vertices.count())
print
print("SAMPLE PRODUCT DATA:")
print(ProductGraph.vertices.take(1))
print

print("Links: %d" % ProductGraph.edges.count())
print
print("SAMPLE LINK DATA:")
print(ProductGraph.edges.take(1))
print

Products: 10611
SAMPLE PRODUCT DATA:
[Row(Id-nu='13260', id='B000002TIR', Group='Music', Salesrank=2004, Reviews=16, Title='French Kiss')]
Links: 52158
SAMPLE LINK DATA:
[Row(src='0790747324', dst='B00007JMD8', link-id=0)]
 Out[ 17 ]: <function print>

The Graph is constructed. Do the fancy analyses!

1) assess the number of in-degress per product and compare it to the salesrank

Plot the next chunk in the following way:

you can look at the distribution of the inDegree-values by clicking on the icon to the right of 'Raw table' and then chosing a histogram, values = 'inDegree' and  30 bins

We can see that most of the inDegree values are between 0 and 2

In [31]:
display(ProductGraph.inDegrees.sort('inDegree', ascending=False))
#inDegRank = ProductGraph.inDegrees.sort('inDegree', ascending=False)

id inDegree B00008LDNZ 78 B00005JMAH 55 B00023P4I8 53 B00003CXCT 52 B000634DCW 45 B000068DBC 37 6304765266 36 B00005ATZT 34 B00005JMIJ 33 B00001QEE7 33 6305368171 33 1594480001 33 B0000TB01Y 32 B00003CXA2 30 B0001NBLVI 29 630522577X 29 B000059HB7 29 B0002CR04A 28 B0001I561E 28 6305696071 27 B00005JMJG 27 0684801523 27 B000022TSH 27 B00003CX95 27 B0002IVN9W 26 B000286RKW 26 B000000XB9 26 0316769487 24 B00005JMQZ 24 B00005JN4W 23 B0000DJZ8R 23 B000065DWB 23 B00005QFDX 23 B00000F14B 22 B00003CXCQ 22 1400032717 22 B00012QM8G 22 B0000026NF 22 B00001QEE6 21 B00005V3Z4 21 B00005Y1XY 21 B0001NBMH6 21 B00005JKGO 21 B0002JP2OI 21 B00005LKHZ 20 B00005LC4A 20 B00004RF97 20 B00000K0DS 20 0790731479 20 B00005KARK 20 B00005YUPQ 20 0446310786 20 B00008975Z 20 0694003611 19 B00005JMQW 19 B00005NTNW 19 B00008CMR4 19 0790729628 19 B000067J1P 19 B0000AOV4I 19 0140177396 19 B00003CXE7 19 B00001MXXH 19 B0002YLCOM 19 6305280770 19 B00004R9A0 18 B00006FDCP 18 B00003CX8Y 18 0805047905 18 B00005JKOI 18 0399226907 18 B00005Q79A 18 B0002I84JO 18 B0002MPQRS 18 0792839749 18 0800141660 18 6305168857 18 B00049QQJQ 18 0399501487 18 B00006RCO1 18 B00000K31Q 18 0066620996 18 B0002VEPRQ 17 B0001L3MI0 17 0449214923 17 B00003CXBU 17 B00003JRCQ 17 B0001Z4OXS 17 0792165055 17 6305144168 17 B00000JS62 17 B00005NQJJ 17 B00005JKDR 16 B000053VBK 16 B0001932ZU 16 B00004XPPG 16 B000055ZF6 16 6305216088 16 B00003CXC7 16 B00005PJ8V 16 B000002ADT 16 B00004U104 16 B00019072G 16 B00004KD24 16 0790729377 16 0440237688 16 B000055Y0X 16 0783225725 16 B000023VTP 16 0064401847 16 6303144535 15 B00004VVNB 15 B0007TKNL0 15 6305212112 15 0783117523 15 B00000DHZJ 15 0440498058 15 1568328478 15 B00003CXD3 15 6304696566 15 B00074DY0M 15 1568322267 15 0792833287 15 B00002CGGH 15 B000035Z1Z 15 B00003CXAT 15 B00005S6K8 15 0553210092 15 B00003CX9E 14 B00005R87R 14 B00005ASGF 14 B00008ZZ7B 14 B00006JMRE 14 B00023P4II 14 6303467814 14 B0000A98ZP 14 B00004RF9B 14 1568328605 14 B00004TDTO 14 B00007J88T 14 B0000058HV 14 B0001XAPZ6 14 0790749815 14 B0000062XS 14 B00000ILEC 14 B0000TG9E2 14 B00005JM02 14 B00005YUPN 14 6305428476 14 B00005QJEF 14 B000056WQX 14 B00005JMUA 14 B00005N5S3 14 630529142X 14 B00005AAA9 14 B00005B6JZ 14 B00007LVCM 14 B00006AL1E 14 0060254920 14 B00009W0WM 14 B00005Q4CS 14 0451526341 14 0064430227 14 B0000BWVA3 14 B00000G3AZ 13 B000002G3Y 13 B00005JN2Z 13 6305428042 13 B0000648WV 13 B00007E2F5 13 B00007GZR1 13 B00008NRI8 13 0783240171 13 079284615X 13 B00005O3VC 13 B0002WZTOI 13 B000065UFD 13 6304907729 13 0312422156 13 B00005MP58 13 B00005PJ8S 13 B000002KAR 13 0783227884 13 B00005LZT9 13 B00001QEE9 13 B00005MLVA 13 6304711891 13 B0006H32DY 13 B0000047CX 13 6304711913 13 B000055XM8 13 B00001QEE8 13 0071381392 13 0689835604 13 B00007JMD8 13 015602943X 13 B00009YXET 13 B00005JKHN 13 B00005221M 13 0780621611 13 B00006IR69 13 0609610570 13 B0000AQS6R 13 0375810455 13 B000000W5P 13 0738920762 13 B00005JN0Q 13 B0000033WI 13 078322592X 13 B00005JSHW 13 B00000I21X 13 B00004Y62W 13 B0000542C9 13 B000087F7D 13 B000002TTP 13 B0000TLA9G 12 B00019330O 12 B00006JSS8 12 0767830555 12 0738920770 12 B00000IQW3 12 B000001AMZ 12 B00004TJKK 12 0790740060 12 630227608X 12 B000067FP3 12 630460291X 12 B00000K3CQ 12 0394800168 12 081297106X 12 B00003RQNJ 12 B00008438T 12 6304696612 12 B00004RF9D 12 B00005JKFR 12 B0000YEEGM 12 6304698526 12 B00003CXDC 12 0800141741 12 B0000BZK1R 12 0440414806 12 0060256656 12 0767803434 12 6304698690 12 B00004R99Q 12 B00005PJ8O 12 B00002RAPT 12 B00003CWL6 12 B00000IBME 12 B000065V3P 12 B000094J7Z 12 B000065U1M 12 B00009NHBM 12 0783233493 12 B00003CXDD 12 B00005N7YZ 12 0345342968 12 B000002KHJ 12 B0006Z2NAY 12 B000002UAO 12 B00005LOKQ 12 B00003CXCZ 12 B0000541AK 12 B0000648WU 12 B0000A02TZ 12 0790731037 12 189176702X 12 B00005JKG1 12 B0002I832C 12 B0000CDL6V 12 B000063K1U 12 B000002OO0 11 B000244EMO 11 B00003CXC5 11 B00006DEF9 11 0792164962 11 B0001FVDGO 11 1568328443 11 B00008WJBE 11 6305310343 11 B00009W5CA 11 B00003CXBK 11 B0000BWVD9 11 B00005QFE1 11 B00005J6UP 1

In [32]:
maxNumInDeg = ProductGraph.inDegrees.groupBy().max('inDegree')
display(maxNumInDeg)

maxNumInDegValue = str(maxNumInDeg.collect()[0]['max(inDegree)'])
#print(maxNumInDegValue)

maxInDegreeProduct = ProductGraph.inDegrees.filter('inDegree = {}'.format(maxNumInDegValue))
display(maxInDegreeProduct)


id inDegree B00008LDNZ 78

In [33]:
# join back to get product info based on max inDegree
ProductMaxInDegree = ProductGraph.vertices.join(maxInDegreeProduct, 'id')
display(ProductMaxInDegree)




id Id-nu Group Salesrank Reviews Title inDegree B00008LDNZ 548091 DVD 110 83 Laura 78

Plot the following chunk in this way:
Plot this as a bar chart with 'Plot Options':
 - Keys: InDegree
 - Values: Salesrank
 
Unfortunately you can't chose to have a logarithmic y-axis, but you do see that at least for low in-degree values the sales rank is rather bad.

Alternative: scatterplot inDegree vs. salesrank and grouping by product group (Keys=Group)

In [35]:
# repeat exercise from above but with entire list
selection = ProductGraph.inDegrees.sort('inDegree', ascending=False)
#display(selection)

ProductSortInDeg = ProductGraph.vertices.join(selection, 'id').sort('inDegree', ascending=False)
display(ProductSortInDeg)

id Id-nu Group Salesrank Reviews Title inDegree B00008LDNZ 548091 DVD 110 83 Laura 78 B000068DBC 546689 DVD 109 633 Pulp Fiction (Collector's Edition) 37 6304765266 97579 DVD 351 152 While You Were Sleeping 36 B00005ATZT 547040 DVD 1020 140 The Fugitive (Special Edition) 34 B00001QEE7 104775 DVD 49 160 The Little Mermaid (Limited Issue) 33 6305368171 276371 DVD 236 500 You've Got Mail 33 B00003CXA2 547272 DVD 185 530 Forrest Gump 30 630522577X 548182 DVD 253 180 My Fair Lady 29 B000059HB7 4388 DVD 1877 79 Rio Bravo 29 B000022TSH 544390 DVD 1108 154 Chinatown 27 6305696071 545472 DVD 540 193 Pretty Woman (10th Anniversary Edition) 27 B00003CX95 166635 DVD 233 745 Braveheart 27 0684801523 199628 Book 956 934 The Great Gatsby 27 B000000XB9 327081 Music 87 163 """Creedence Clearwater Revival - Chronicle, Vol. 1: The 20 Greatest Hits""" 26 0316769487 98756 Book 60 2568 The Catcher in the Rye 24 B000065DWB 46130 Music 417 31 Foreigner - Complete Greatest Hits 23 B00005QFDX 220569 DVD 330 115 Sesame Street - The Best of Elmo 23 B00000F14B 216263 DVD 351 123 The Music Man (Special Edition) 22 B0000026NF 3499 Music 153 180 Journey - Greatest Hits 22 B00003CXCQ 402685 DVD 679 247 Snow White and the Seven Dwarfs (Disney Special Platinum Edition) 22 B00001QEE6 444195 DVD 904 68 Lady and the Tramp (Limited Issue) 21 B00005V3Z4 546221 DVD 176 848 Donnie Darko 21 B00005Y1XY 429734 Music 333 47 Cars - Complete Greatest Hits 21 B00004RF97 96590 DVD 792 104 The Philadelphia Story 20 B00005KARK 453361 DVD 353 140 Dumbo (60th Anniversary Edition) 20 B00005LKHZ 164067 DVD 343 236 The NeverEnding Story 20 B00005LC4A 545385 DVD 1260 158 Charade 20 0446310786 537519 Book 111 1414 To Kill a Mockingbird 20 0790731479 138550 DVD 633 126 Beetlejuice 20 B00008975Z 270909 DVD 695 407 Reservoir Dogs 20 B00005YUPQ 28804 DVD 389 129 Baby Einstein - Baby Van Gogh - World of Colors 20 B00003CXE7 520382 DVD 1702 1513 Gladiator 19 B00008CMR4 547333 DVD 1798 38 Big Jake 19 B00001MXXH 548318 DVD 652 295 Ferris Bueller's Day Off 19 0140177396 341570 Book 126 1000 Of Mice and Men (Penguin Great Books of the 20th Century) 19 0694003611 502086 Book 156 339 Goodnight Moon (Board Book) 19 0790729628 545235 DVD 392 812 Blade Runner (The Director's Cut) 19 6305280770 546760 DVD 1915 79 The King and I 19 B000067J1P 545271 DVD 1331 357 The Sound of Music (Single Disc Full Screen Edition) 19 B00005NTNW 548266 DVD 768 78 The Outlaw Josey Wales 19 0805047905 502784 Book 171 172 Brown Bear, Brown Bear, What Do You See? 18 B00003CX8Y 191667 DVD 439 354 Beauty and the Beast (Disney Special Platinum Edition) 18 0399226907 180888 Book 279 164 The Very Hungry Caterpillar board book 18 0399501487 35512 Book 143 1101 Lord of the Flies 18 B00005JKOI 137583 DVD 2582 75 Pretty in Pink 18 6305168857 152642 DVD 154 234 Young Frankenstein (Special Edition) 18 B00006FDCP 59459 DVD 1582 224 One Flew Over the Cuckoo's Nest (Two-Disc Special Edition) 18 B00006RCO1 544551 DVD 979 162 All About Eve (Special Edition) 18 B00005Q79A 191370 DVD 1169 166 What's Eating Gilbert Grape 18 0800141660 546277 DVD 443 250 Sense and Sensibility 18 0066620996 154855 Book 29 363 Good to Great: Why Some Companies Make the Leap... and Others Don't 18 0792839749 12389 DVD 384 125 Chitty Chitty Bang Bang (Full Screen Edition) 18 B00000K31Q 18824 DVD 1379 387 10 Things I Hate About You 18 B00003CXBU 543078 DVD 1278 119 On the Waterfront (Special Edition) 17 B00005NQJJ 469947 Music 257 26 Doobie Brothers - Greatest Hits 17 0449214923 482263 Book 215 260 Think and Grow Rich 17 6305144168 544508 DVD 972 276 Dead Poets Society 17 B00003JRCQ 424377 DVD 462 204 Persuasion 17 B00000JS62 547782 DVD 55 380 The Wizard of Oz 17 0792165055 544654 DVD 1993 163 Romeo & Juliet 17 0783225725 547690 DVD 3034 64 High Plains Drifter 16 B000002ADT 484602 Music 117 541 Kind of Blue 16 B000055ZF6 546103 DVD 3509 57 Terms of Endearment 16 B000023VTP 547943 DVD 505 329 Notting Hill (Collector's Edition) 16 B000055Y0X 235687 DVD 345 311 Dr. S

Plot inDegree vs. Salesrank for Books only
* plot as scatter plot with LOESS smoother

In [37]:
# try to differentiate by product group
ProductSortInDeg = ProductGraph.vertices.join(selection, 'id').sort('inDegree', ascending=False).filter("Group = 'Book'")
display(ProductSortInDeg)

id Id-nu Group Salesrank Reviews Title inDegree 0684801523 199628 Book 956 934 The Great Gatsby 27 0316769487 98756 Book 60 2568 The Catcher in the Rye 24 0446310786 537519 Book 111 1414 To Kill a Mockingbird 20 0694003611 502086 Book 156 339 Goodnight Moon (Board Book) 19 0140177396 341570 Book 126 1000 Of Mice and Men (Penguin Great Books of the 20th Century) 19 0066620996 154855 Book 29 363 Good to Great: Why Some Companies Make the Leap... and Others Don't 18 0805047905 502784 Book 171 172 Brown Bear, Brown Bear, What Do You See? 18 0399501487 35512 Book 143 1101 Lord of the Flies 18 0399226907 180888 Book 279 164 The Very Hungry Caterpillar board book 18 0449214923 482263 Book 215 260 Think and Grow Rich 17 0440237688 188765 Book 428 2617 The Giver 16 0064401847 42948 Book 1258 655 Bridge to Terabithia 16 0553210092 317053 Book 368 348 The Scarlet Letter 15 0440498058 291216 Book 414 872 A Wrinkle in Time 15 0451526341 53175 Book 275 990 Animal Farm 14 0064430227 206514 Book 1032 76 Harold and the Purple Crayon 50th Anniversary Edition (Purple Crayon Books) 14 0060254920 268369 Book 148 211 Where the Wild Things Are 14 0689835604 335281 Book 250 90 Where Is Baby's Belly Button? 13 0609610570 429788 Book 143 157 Execution: The Discipline of Getting Things Done 13 0375810455 220181 Book 209 20 Too Big for Diapers (Too Big Board Books) 13 0312422156 425735 Book 133 517 Middlesex: A Novel 13 0071381392 18173 Book 239 355 The No-Cry Sleep Solution: Gentle Ways to Help Your Baby Sleep Through the Night 13 0060256656 239578 Book 360 394 The Giving Tree 12 0440414806 528497 Book 761 2922 Holes (Yearling Newbery) 12 189176702X 532007 Book 1725 8 "How to be the Leader of the Pack...And have Your Dog Love You For It. (""How to"" booklets from Dog's Best Friend)" 12 0394800168 261029 Book 895 148 Green Eggs and Ham (I Can Read It All by Myself Beginner Books) 12 0345342968 29543 Book 81 1069 Fahrenheit 451 12 0439064864 216201 Book 445 2422 Harry Potter and the Chamber of Secrets (Book 2) 11 043935806X 429348 Book 110 5545 Harry Potter and the Order of the Phoenix (Book 5) 11 0671723650 53885 Book 64 377 How to Win Friends & Influence People 10 0486291138 515301 Book 290 35 1001 Most Useful Spanish Words (Beginners' Guides) 10 0446677450 162283 Book 37 1637 Rich Dad, Poor Dad: What the Rich Teach Their Kids About Money--That the Poor and Middle Class Do Not! 10 0787960756 33690 Book 89 61 The Five Dysfunctions of a Team: A Leadership Fable 10 0486280861 296016 Book 122 13 Easy Spanish Phrase Book: Over 770 Basic Phrases for Everyday Use 10 1573441449 436250 Book 467 32 The Ultimate Guide to Cunnilingus: How to Go Down on a Woman and Give Her Exquisite Pleasure (Ultimate Everything!!!) 10 0440413281 40399 Book 930 350 Bud, Not Buddy (Newbery Medal Winner, 2000) 10 0345446798 240297 Book 2251 77 The Other End of the Leash 10 076360013X 33247 Book 295 209 Guess How Much I Love You (Guess How Much I Love You) 10 0440136482 532042 Book 161 351 Holy Blood, Holy Grail 10 0142001740 337971 Book 26 996 The Secret Life of Bees 10 0439136350 404634 Book 356 2396 Harry Potter and the Prisoner of Azkaban (Book 3) 9 0689711816 179199 Book 1414 251 From the Mixed-Up Files of Mrs. Basil E. Frankweiler 9 0789205386 62274 Book 353 158 The Expectant Father: Facts, Tips and Advice for Dads-To-Be 9 0684852861 357720 Book 215 189 First, Break All the Rules: What the World's Greatest Managers Do Differently 9 0446677469 234402 Book 384 155 Rich Dad's Guide to Investing: What the Rich Invest in, That the Poor and the Middle Class Do Not! 9 0440227534 492127 Book 695 548 Number the Stars (Laurel Leaf Books) 9 0671708635 28211 Book 405 583 The 7 Habits of Highly Effective People 9 0440439884 105186 Book 1457 400 Island of the Blue Dolphins 9 0142000663 211438 Book 201 517 The Grapes of Wrath : (Centennial Edition) 9 0380811960 390380 Book 275 98 How to Talk So Kids Will Listen & Listen So Kids Will Talk 8 1880685000 475853 Book 72 55 The Theory of Poker 8 044

Plot the following chunk in this way:
Plot this as a bar chart with 'Plot Options':
 - Keys: OutDegree
 - Values: Salesrank

In [39]:
# repeat exercise from above for outDegree
selection = ProductGraph.outDegrees.sort('outDegree', ascending=True)
#display(selection)

ProductSortOutDeg = ProductGraph.vertices.join(selection, 'id').sort('OutDegree', ascending=True)
display(ProductSortOutDeg)

outDegree is rather boring, since most of the products get assigned 5 products by default that are presumingly similar, as you can see here:

In [41]:
display(ProductSortOutDeg.groupBy("outDegree").count())

plot the output of the next chunk in the following way:
chose a bar chart with options:
  - keys: Reviews
  - values: Salesrank
  
and apply to all rows.
We can see that the less reviews a product has, the less likely it is to be a top selling product (the causal orientation might also be inverse here :) )

In [43]:
# relation of number of reviews and sales rank:
display(ProductGraph.vertices.select("Reviews", "Salesrank").sort("Reviews", ascending = False))


trying to find products which have an edge to another product which in turn has an edge to the first product:

In [45]:
ProductsWithBalancedSimilarity = ProductGraph.find("(v1)-[e1]->(v2); (v2)-[e2]->(v1)")
display(ProductsWithBalancedSimilarity)

how many product-pairs reference each other? can we just calculate it like this?

In [47]:
ProductsWithBalancedSimilarity.count()

does this mean, we have 11202 product-pairs that reference each other? Let's test if our data frame of mutually referencing products lists the pairs once or twice by looking at an example:

In [49]:
display(ProductsWithBalancedSimilarity.filter("(v1.id = '0790747324' and v2.id = 'B00007JMD8') or (v2.id = '0790747324' and v1.id = 'B00007JMD8')"))

So the pairs are listed twice, which kind of makes sense, seeing how we constructed the data frame. This means there are 11202/2 = 5601 product-pairs that reference each other.

Now we want to explore if the "strongly connected components" of our graph are evenly distributed across all groups, which would mean that the probability of a DVD referencing to another DVD is the same as it referencing to e.g. a book, or if the groups tend to stick to themselves.

Here we first calculate the strongly connected components
(small warning: this command takes quite a while, ~1.17 hours, 238 jobs):

In [52]:
connectedComp = ProductGraph.stronglyConnectedComponents(maxIter=3)
#connectedComp.select("id", "component").groupBy("component").count().sort(desc("count")).show()

#ProductGraph.connectedComponents()
display(connectedComp)

The connected components are marked by the same value in the column "component". So if we group the data frame by "component" and "Group" and use the function "count()", we will have as many rows per "component" value as the number of different groups belonging to this cluster.

In [54]:
countGroupComponent= connectedComp.groupBy("component", "Group").count().sort("component")
display(countGroupComponent)

Now how many groups are taking part in each cluster?

In [56]:
countComponentsNumberOfGroups = countGroupComponent.groupBy("component").count().sort(desc("count"))
display(countComponentsNumberOfGroups)

Next we get an overview how many clusters there are with members of *n* groups:

In [58]:
numGroupsCount = countComponentsNumberOfGroups.withColumnRenamed("count", "NumberOfGroupsForComponent").groupBy("NumberOfGroupsForComponent").count()
display(numGroupsCount)

So how many clusters are there in total? Just adding up the values in column "count" in the last chunk results in the number of clusters:

In [60]:
display(numGroupsCount.agg({"count": "sum"}))

6751 is a very high number, seeing that there are just 10611 products in our dataset in total.

And what is the size distribution of the clusters in general (independent which groups its members belong to)?

In [63]:
countConnected = connectedComp.groupBy("component").count().sort(desc("count"))
display(countConnected)

So there is only one really huge cluster of strongly connected components with almost 1000 members. The next biggest one has only a size of 29.

How many of the 6751 clusters have more than 3 members?

In [66]:
countConnected.filter("count > 3").count()